### ISSUE TITLE
DDB assert when using TCP connections in app

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#472]
The database dies when a client connects to my Acton application that is a TCP listening server.

```
kll@Boxy:~/terastream/acton/test$ ../dist/bin/actondb 
SERVER: Using args: portno=32000, gportno=34000, verbosity=1, seeds:
Test create ACTORS_TABLE - OK (0)
Test create MSGS_TABLE - OK (0)
Test create_state_schema - OK
BACKEND: Queue table 2 created
Test create MSG_QUEUE table - OK (0)
Test create_queue_schema - OK
SERVER: Started [0.0.0.0:32000, 127.0.0.1:34000], my_lc = VC(99536:0, )
SERVER: accepted connection from client: localhost (127.0.0.1:37334)
BACKEND: Queue 2/0 created
BACKEND: Subscriber 0/0/0 subscribed queue 2/0 with callback (nil)
BACKEND: Queue 2/-11 created
BACKEND: Subscriber 0/0/-11 subscribed queue 2/-11 with callback (nil)
BACKEND: Queue 2/-12 created
BACKEND: Subscriber 0/0/-12 subscribed queue 2/-12 with callback (nil)
BACKEND: Inserted queue entry 0 in queue 2/-12, status=0
BACKEND: Notified remote subscriber -12
BACKEND: Queue 2/-14 created
BACKEND: Subscriber 0/0/-14 subscribed queue 2/-14 with callback (nil)
BACKEND: Inserted queue entry 0 in queue 2/-14, status=0
BACKEND: Notified remote subscriber -14
BACKEND: Subscriber -14 read 1 queue entries, new_read_head=0
BACKEND: Inserted queue entry 0 in queue 2/-11, status=0
BACKEND: Notified remote subscriber -11
BACKEND: Subscriber -11 read 1 queue entries, new_read_head=0
BACKEND: Subscriber -12 read 1 queue entries, new_read_head=0
actondb: backend/txns.c:498: persist_txn: Assertion `res == 0' failed.
Aborted
```

```
kll@Boxy:~/terastream/acton/examples$ ./server --rts-verbose --rts-debug --rts-ddb-host localhost --rts-ddb-replication 1
#RTS# Detected 32 CPUs: Using 32 worker threads for 1:1 mapping with CPU affinity set.
#RTS# Using distributed database backend replication factor of 1
#RTS# Using distributed database backend (DDB): localhost:32000
#RTS# Checking for existing actor state in DDB... done
#RTS# No previous state in DDB; Initializing database...#RTS# #### Create queue 0 returns 0
#RTS#    # Subscribe queue 0 returns 0
#RTS# # New Actor -11 at 0x847da0 of class $Env
#RTS# # New Actor -12 at 0x847b70 of class Actor
#RTS# #### Create queue -11 returns 0
#RTS#    # Subscribe queue -11 returns 0
#RTS# #### Create queue -12 returns 0
#RTS#    # Subscribe queue -12 returns 0
#RTS# #### Serializing Actor -11
#RTS# ---  0: class_id    126, blob_size:   0, blob: .
#RTS# ---  1: class_id      0, blob_size:   0, blob: .
#RTS# ---  2: class_id     13, blob_size:   1, blob: 0 .
#RTS# ---  3: class_id      0, blob_size:   0, blob: .
#RTS# ---  4: class_id      7, blob_size:   1, blob: 2 .
#RTS# ---  5: class_id      6, blob_size:   4, blob: 8 8 8243124926671957806 0 .
#RTS# ---  6: class_id      6, blob_size:   3, blob: 4 4 892547633 .
#RTS# ## Built blob, size: 16, blob: 126 0 4294967309 0 0 4294967303 2 17179869190 8 8 8243124926671957806 0 12884901894 4 4 892547633 .
#RTS#    # insert to table 0, row -11, returns 0
#RTS#    # enqueue bootstrap msg -13 to ancestor0 queue -12 returns 0
done
#RTS# ## Running actor -12 : Actor
#RTS# # New Actor -14 at 0x7f9e90000b60 of class server$$main
#RTS# #### Create queue -14 returns 0
#RTS#    # Subscribe queue -14 returns 0
#RTS# #### Serializing Actor -12
#RTS# ---  0: class_id     15, blob_size:   0, blob: .
#RTS# ---  1: class_id      0, blob_size:   0, blob: .
#RTS# ---  2: class_id     13, blob_size:   1, blob: 0 .
#RTS# ---  3: class_id      0, blob_size:   0, blob: .
#RTS# ## Built blob, size: 5, blob: 15 0 4294967309 0 0 .
#RTS#    # insert to table 0, row -12, returns 0
#RTS# #### Serializing Msg -15
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  2: class_id    135, blob_size:   0, blob: .
#RTS# ---  3: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  4: class_id   -126, blob_size:   1, blob: -11 .
#RTS# ---  5: class_id     13, blob_size:   1, blob: 1642427791965282 .
#RTS# ---  6: class_id    123, blob_size:   0, blob: .
#RTS# ## Built blob, size: 11, blob: 14 8589934455 18446744073709551602 135 8589934455 18446744073709551602 8589934466 18446744073709551605 4294967309 1642427791965282 123 .
#RTS#    # insert to table 1, row -15, returns 0
#RTS# #### FLUSH_outgoing messages from -12
#RTS#    # enqueue msg -15 to queue -14 returns 0
#RTS# #### Serializing Msg -13
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id    -15, blob_size:   1, blob: -12 .
#RTS# ---  2: class_id    124, blob_size:   0, blob: .
#RTS# ---  3: class_id     13, blob_size:   1, blob: 1642427791965282 .
#RTS# ---  4: class_id    125, blob_size:   0, blob: .
#RTS# ## Built blob, size: 7, blob: 14 8589934577 18446744073709551604 124 4294967309 1642427791965282 125 .
#RTS#    # insert to table 1, row -13, returns 0
#RTS# ## Running actor -14 : server$$main
#RTS# ############## Commit

#RTS# ## AWAIT actor -12 : Actor
#RTS# ## CONT actor -14 : server$$main
#RTS# ## Running actor -14 : server$$main
#RTS# #### Serializing Actor -14
#RTS# ---  0: class_id    137, blob_size:   0, blob: .
#RTS# ---  1: class_id      0, blob_size:   0, blob: .
#RTS# ---  2: class_id     13, blob_size:   1, blob: 1 .
#RTS# ---  3: class_id      0, blob_size:   0, blob: .
#RTS# ---  4: class_id   -126, blob_size:   1, blob: -11 .
#RTS# ---  5: class_id     80, blob_size:   0, blob: .
#RTS# ---  6: class_id     81, blob_size:   0, blob: .
#RTS# ---  7: class_id    -80, blob_size:   1, blob: 5 .
#RTS# ---  8: class_id     82, blob_size:   0, blob: .
#RTS# ---  9: class_id    -80, blob_size:   1, blob: 5 .
#RTS# ## Built blob, size: 14, blob: 137 0 4294967309 1 0 8589934466 18446744073709551605 80 81 8589934512 5 82 8589934512 5 .
#RTS#    # insert to table 0, row -14, returns 0
#RTS# #### Serializing Msg -16
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id   -126, blob_size:   1, blob: -11 .
#RTS# ---  2: class_id    108, blob_size:   0, blob: .
#RTS# ---  3: class_id   -126, blob_size:   1, blob: -11 .
#RTS# ---  4: class_id      2, blob_size:   1, blob: 1235 .
#RTS# ---  5: class_id    129, blob_size:   0, blob: .
#RTS# ---  6: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  7: class_id     13, blob_size:   1, blob: 1642427791965282 .
#RTS# ---  8: class_id    123, blob_size:   0, blob: .
#RTS# ## Built blob, size: 14, blob: 14 8589934466 18446744073709551605 108 8589934466 18446744073709551605 4294967298 1235 129 8589934455 18446744073709551602 4294967309 1642427791965282 123 .
#RTS#    # insert to table 1, row -16, returns 0
#RTS# #### FLUSH_outgoing messages from -14
#RTS# ## Running actor -11 : $Env
#RTS#    # enqueue msg -16 to queue -11 returns 0
#RTS# #### Serializing Msg -15
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  2: class_id    123, blob_size:   0, blob: .
#RTS# ---  3: class_id     13, blob_size:   1, blob: 1642427791965282 .
#RTS# ---  4: class_id      0, blob_size:   0, blob: .
#RTS# ## Built blob, size: 7, blob: 14 8589934455 18446744073709551602 123 4294967309 1642427791965282 0 .
#RTS# ## CONT actor -11 : $Env
#RTS# ## Running actor -11 : $Env
#RTS#    # insert to table 1, row -15, returns 0
#RTS# #### Serializing Actor -11
#RTS# ---  0: class_id    126, blob_size:   0, blob: .
#RTS# ---  1: class_id      0, blob_size:   0, blob: .
#RTS# ---  2: class_id     13, blob_size:   1, blob: 1 .
#RTS# ---  3: class_id      0, blob_size:   0, blob: .
#RTS# ---  4: class_id      7, blob_size:   1, blob: 2 #RTS#    # dummy read msg from queue -14 returns 1, entries read: 1
.
#RTS# ---  5: class_id      6, blob_size:   4, blob: 8 8 8243124926671957806 0 .
#RTS# ---  6: class_id      6, blob_size:   3, blob: 4 4 892547633 .
#RTS# ## Built blob, size: 16, blob: 126 0 4294967309 1 0 4294967303 2 17179869190 8 8 8243124926671957806 0 12884901894 4 4 892547633 .
#RTS#    # consume msg -15 from queue -14 returns 0
#RTS#    # insert to table 0, row -11, returns 0
#RTS# #### FLUSH_outgoing messages from -11
#RTS# #### Serializing Msg -16
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id   -126, blob_size:   1, blob: -11 .
#RTS# ---  2: class_id    123, blob_size:   0, blob: .
#RTS# ---  3: class_id     13, blob_size:   1, blob: 1642427791965282 .
#RTS# ---  4: class_id      0, blob_size:   0, blob: .
#RTS# ## Built blob, size: 7, blob: 14 8589934466 18446744073709551605 123 4294967309 1642427791965282 0 .
#RTS# ############## Commit

#RTS# ## Waking up actor -12 : Actor
#RTS# ## DONE actor -14 : server$$main
#RTS# ## Running actor -12 : Actor
#RTS# ## CONT actor -12 : Actor
#RTS# ## Running actor -12 : Actor
#RTS#    # insert to table 1, row -16, returns 0
#RTS# #### Serializing Actor -12
#RTS# ---  0: class_id     15, blob_size:   0, blob: .
#RTS# ---  1: class_id      0, blob_size:   0, blob: .
#RTS# ---  2: class_id     13, blob_size:   1, blob: 1 .
#RTS# ---  3: class_id      0, blob_size:   0, blob: .
#RTS# ## Built blob, size: 5, blob: 15 0 4294967309 1 0 .
#RTS#    # dummy read msg from queue -11 returns 1, entries read: 1
#RTS#    # insert to table 0, row -12, returns 0
#RTS# #### FLUSH_outgoing messages from -12
#RTS# #### Serializing Msg -13
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id    -15, blob_size:   1, blob: -12 .
#RTS# ---  2: class_id    125, blob_size:   0, blob: .
#RTS# ---  3: class_id     13, blob_size:   1, blob: 1642427791965282 .
#RTS# ---  4: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ## Built blob, size: 8, blob: 14 8589934577 18446744073709551604 125 4294967309 1642427791965282 8589934455 18446744073709551602 .
#RTS#    # consume msg -16 from queue -11 returns 0
#RTS#    # insert to table 1, row -13, returns 0
#RTS#    # dummy read msg from queue -12 returns 1, entries read: 1
#RTS# ############## Commit

#RTS# ## DONE actor -11 : $Env
#RTS#    # consume msg -13 from queue -12 returns 0
#RTS# ############## Commit

#RTS# ## DONE actor -12 : Actor
```
This is as far as we process at first, then I connect a client using
```
nc localhost 12345
```

and then Acton continues... and crashes...

```
#RTS# # New Actor -17 at 0x7f9e74000b60 of class $Connection
#RTS# ## Running actor -14 : server$$main
Connection from localhost
#RTS# ## CONT actor -14 : server$$main
#RTS# ## Running actor -14 : server$$main
#RTS# #### Serializing Actor -14
#RTS# ---  0: class_id    137, blob_size:   0, blob: .
#RTS# ---  1: class_id      0, blob_size:   0, blob: .
#RTS# ---  2: class_id     13, blob_size:   1, blob: 2 .
#RTS# ---  3: class_id      0, blob_size:   0, blob: .
#RTS# ---  4: class_id   -126, blob_size:   1, blob: -11 .
#RTS# ---  5: class_id     80, blob_size:   0, blob: .
#RTS# ---  6: class_id     81, blob_size:   0, blob: .
#RTS# ---  7: class_id    -80, blob_size:   1, blob: 5 .
#RTS# ---  8: class_id     82, blob_size:   0, blob: .
#RTS# ---  9: class_id    -80, blob_size:   1, blob: 5 .
#RTS# ## Built blob, size: 14, blob: 137 0 4294967309 2 0 8589934466 18446744073709551605 80 81 8589934512 5 82 8589934512 5 .
#RTS#    # insert to table 0, row -14, returns 0
#RTS# #### Serializing Msg -19
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id   -127, blob_size:   1, blob: -17 .
#RTS# ---  2: class_id    114, blob_size:   0, blob: .
#RTS# ---  3: class_id   -127, blob_size:   1, blob: -17 .
#RTS# ---  4: class_id    131, blob_size:   0, blob: .
#RTS# ---  5: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  6: class_id    133, blob_size:   0, blob: .
#RTS# ---  7: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  8: class_id     13, blob_size:   1, blob: 1642427796379172 .
#RTS# ---  9: class_id    123, blob_size:   0, blob: .
#RTS# ## Built blob, size: 15, blob: 14 8589934465 18446744073709551599 114 8589934465 18446744073709551599 131 8589934455 18446744073709551602 133 8589934455 18446744073709551602 4294967309 1642427796379172 123 .
#RTS#    # insert to table 1, row -19, returns 0
#RTS# #### FLUSH_outgoing messages from -14
#RTS#    # enqueue msg -19 to queue -17 returns 0
#RTS# #### Serializing Msg -18
#RTS# ---  0: class_id     14, blob_size:   0, blob: .
#RTS# ---  1: class_id   -137, blob_size:   1, blob: -14 .
#RTS# ---  2: class_id    123, blob_size:   0, blob: .
#RTS# ---  3: class_id     13, blob_size:   1, blob: 1642427796379172 .
#RTS# ---  4: class_id      0, blob_size:   0, blob: .
#RTS# ## Built blob, size: 7, blob: 14 8589934455 18446744073709551602 123 4294967309 1642427796379172 0 .
#RTS#    # insert to table 1, row -18, returns 0
#RTS#    # dummy read msg from queue -14 returns 1, entries read: 0
#RTS#    # consume msg -18 from queue -14 returns 0
Host disconnected , ip 127.0.0.1 , port 32000 
No quorum (0/1 replies received)                                                                                                             22cÕ°C¦욖½Ks8£退ð¸Љ
Љ
No quorum (0/1 replies received)                                                                                                             22cÕ°C¦욖½Ks8?ðYЉ
```

Not sure whether the RTS tries to use the database incorrectly or if there is an actual bug in the database. At the very least, I don't want the database to die - IMHO, if there's some form of incorrect use, it should reject / abort the transaction but continue processing. Thus, I'm starting the classification as a DB issue here.

**Comment by plajjan** [comment]
Enabling VERBOSE_TXNS_PERSIST and printing the res I see:


```
BACKEND: Txn 7024d523-562a-4e8b-a06e-a885ee4dca20 successfully persisted write of type 9 with res -2
```

type 9 is ENQUEUE, and -2 is DB_ERR_NO_QUEUE.

The last debug output from RTS is "consume msg...".

Looking in rts.c:
```
            int ret = remote_consume_queue_in_txn(($WORD)key, 0, 0, MSG_QUEUE, ($WORD)key, read_head, txnid, db);
            rtsd_printf(LOGPFX "   # consume msg %ld from TIMER_QUEUE returns %d\n", m->$globkey, ret);
            int ret2 = remote_enqueue_in_txn(($WORD*)&m->$globkey, 1, NULL, 0, MSG_QUEUE, (WORD)m->$to->$globkey, txnid, db);
            rtsd_printf(LOGPFX "   # (timed) enqueue msg %ld to queue %ld returns %d\n", m->$globkey, m->$to->$globkey, ret2);
```

So the thing we do after that message is remote_enqueue_in_txn, presumably to a queue that does not exist.

It seems we now muddle the result, like persist_write calls a bunch of different functions depending on the query_type. If all those called functions returned according to the same list of errors, we could just pass that error on and actually make sense of it on the client side. @aagapi WDYT?

**Comment by plajjan** [comment]
Fixed through #481 so going to close. I will open another issue about DB error handling.
